mypy/join.py

```diff
@@ -137,7 +140,16 @@ def visit_type_var(self, t: TypeVarType) -> Type:
 
     def visit_instance(self, t: Instance) -> Type:
         if isinstance(self.s, Instance):
-            return join_instances(t, self.s)
+            nominal = join_instances(t, self.s)
+            structural = None  # type: Optional[Instance]
+            if t.type.is_protocol and is_protocol_implementation(self.s, t):
+                structural = t
+            elif self.s.type.is_protocol and is_protocol_implementation(t, self.s):
+                structural = self.s
+            # structural type for join is preferred
```

Comment:
Add comment about when a nominal join is preferred. 

Style nit: use capital 's' for 'structural'.